---
title: "DEGsCont"
author: "Ying"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r package loading}
library(tidyverse)
library(magrittr)
library(edgeR)
library(ComplexHeatmap) # for UpSet Plot

```

```{r DGEList}
# import data and choose cont condition samples
geneCounts <- 
  read_delim("../Ying_Submergence_batch1_final.txt", delim = "\t", comment = "#") %>% 
  column_to_rownames("Geneid") %>% 
  dplyr::select(contains("cont"))
colnames(geneCounts) %<>% 
  str_remove("../3_BAM/") %>% 
  str_remove(".STARAligned_TAIR10.sortedByCoord.bam")

# create meta data with ordered 'group' levels
meta <- 
  tibble(sample = colnames(geneCounts),
         group = str_extract(sample, ".+(?=-cont)"), # set genotypes as 'group', 7 levels
         reps = str_extract(sample, "(?<=cont-).+"))
genoLevels <- c("Col", "gad1CS", "gad1SALK", "gad21", "gad1245", "gad2OEB33", "pop28")
meta$group <- factor(meta$group, levels = genoLevels)

genoColour <- 
  c("#F8766D", "#7CAE00", "#00BE67", "#00BFC4", "#CD9600", "#C77CFF", "#FF61CC")
names(genoColour) <- genoLevels

# DGEList, filtering, Normalization (RNA composition)
countList <- 
  DGEList(counts = geneCounts, samples = meta)
keep <- filterByExpr(countList)
table(keep)
countList <- countList[keep,,keep.lib.sizes=FALSE] 
countList <- countList %>% calcNormFactors()

# set WT Col-cont as reference
countList$samples$group <- 
  factor(countList$samples$group, levels = genoLevels)
```

```{r mds plot}
# visualize sample clustering using MDS plot
# "pairwise" to choose the top genes separately for each pairwise comparison between the samples
PCoA <-
  plotMDS(countList, col=as.numeric(countList$samples$group),
          method = "logFC", 
          gene.selection = "pairwise"  # the default is "pairwise" for PCoA plot
          )

PCoA_plot <- PCoA %>% 
  extract(c(5,9,10)) %>% # coordinates info
  as.data.frame() %>% 
  rownames_to_column("sample") %>% 
  left_join(meta) %>% 
  ggplot(aes(x = x, y = y)) +
  geom_point(aes(colour = group), size = 4, alpha = 1) +
  geom_polygon(aes(fill = group), alpha = 0.4) +
  theme_bw() +
  theme(axis.text.x = element_text(vjust = -2),
        legend.position = c(0.5, 0.2),
        legend.margin = margin(-14, 0,0,0),
        text = element_text(size = 15),) +
  labs(x = "Leading logFC dim 1 (48%)",
       y = "Leading logFC dim 2 (13%)"
       ) +
  scale_fill_manual(values = genoColour) +
  scale_color_manual(values = genoColour)
```

```{r estimating dispersions}
# if with multiple factors, dispersion needs to be estimated with specified Design Matrix (see section 2.10.2 in edgeR user's guide updated in 21 Oct. 2019)
counts_Disp <- estimateDisp(countList)
plotBCV(counts_Disp)
```

```{r design matrix, model fitting}
# construct design matrix in the ordered levels of genotypes i.e. levels of countList$sample$group

designMat0 <- model.matrix(~ 0 + group, data = countList$samples) # no ref-group
colnames(designMat0) <- levels(countList$samples$group)

# fit to model: use glmQLFit (to estimate) and plotQLDisp (to visualize) QL dispersion 
fit0 <- 
  glmQLFit(counts_Disp, design = designMat0) 

plotQLDisp(fit0)

```

```{r DE genes}
# to compare gad1CS vs wt, gad1SALK vs wt, etc.
#     "Col", "gad1CS", "gad1SALK", "gad21", "gad1245", "gad2OEB33", "pop28")
# coef = 1,   2,        3,          4,      5,          6,           7

# make contrasts 
myContrasts <-   # a lists of 6 pairs of comparisons
  makeContrasts(
    gad1CS = gad1CS - Col, 
    gad1SALK = gad1SALK - Col, 
    'gad2-1' = gad21 - Col, 
    gad1245 = gad1245 - Col, 
    gad2OE_B33 = gad2OEB33 - Col, 
    'pop2-8' = pop28 - Col, 
    levels = designMat0) # same results using either designMat or designMat0

# GeneID to rowname and add a column of adjust.p-value, then plot DEGs
DEGs <- lapply(colnames(myContrasts), function(c){
  qlf <- glmQLFTest(fit0, contrast = myContrasts[, c])
  qlf$table %>% rownames_to_column("GeneID") %>% 
    mutate(adj.p = p.adjust(PValue, method = "BH"),
           MutantVsWt = c)
})
names(DEGs) <- colnames(myContrasts)

# volcano plot (facetwrap 3 by 2) and DEGs in colour
lapply(names(DEGs), function(g){  # g: genotype_WT, the comparison
  DEGs[[g]] %>% dplyr::select(logFC, adj.p, MutantVsWt) %>% 
    mutate(Expr = ifelse(adj.p < 0.05 & abs(logFC) >=1, 
                         ifelse(logFC > 1, "Up", "Down"),
                         "NotSig"))}) %>% bind_rows() %>% as.data.frame() %>% 
  rstatix::reorder_levels("MutantVsWt", order = names(DEGs)) %>% 
    ggplot(aes(x = logFC,
               y = -log10(adj.p),
               colour = ordered(Expr, 
                                levels = c("Up", "Down", "NotSig"))))+
    geom_point(alpha = 0.2, size = 4, shape=20) +
    scale_color_manual(values = c("red", "blue", "grey")) +
    geom_vline(xintercept = c(-1, 1),
               lty = 4, colour = "#4a452a", lwd = 0.4) +
    geom_hline(yintercept = -log10(0.05),
               lty = 4, colour = "#4a452a", lwd = 0.4) +
    labs(x = "log2 (fold change)",
         y = "- log10 (adjust.p-value)") +
    theme_bw() +
    theme(legend.position = c(0.6,0.92),
          legend.title = element_blank()) +
  facet_wrap(vars(MutantVsWt), nrow = 3, scales = "free")

```

```{r DEGs for go analysis}

# Significant genes in DEGs to do the next enrichment analysis
genes2go <- lapply(names(DEGs), function(g){  # g: each genotype of mutant VS WT
  DEGs[[g]] %>% 
    filter(adj.p < 0.05 & abs(logFC) > 1) %>% 
    mutate(Expr = ifelse(logFC > 1, "Up", "Down"))
}) 
names(genes2go) <- names(DEGs)

# count how many genes in genes2go are up/down regulated genes after filtering
countgenes2go <- lapply(names(DEGs), function(g){
  Up = as.numeric(sum(genes2go[[g]]$Expr == "Up"))
  Down = as.numeric(sum(genes2go[[g]]$Expr == "Down"))
  Total = Up+Down
  Mutant = g
  df <- data.frame(Mutant, Up, Down, Total)
}) %>% 
  bind_rows()

```

```{r UpSet plot for DEGs to go}
# get filterd DEGs (from genes2go)
geneList <- lapply(names(DEGs),function(g){
  genes2go[[g]]$GeneID 
}) 
names(geneList) <- names(DEGs)

# calculate total DEGs among 6 groups
length(geneList %>% unlist() %>% unique()) # 3494

#
# upset plot ------
m0 <- make_comb_mat(geneList) 
m <- m0[comb_size(m0) > 30]

ss <- set_size(m)
cs <- comb_size(m)
od = order(comb_degree(m), -cs)
UpSet(m, 
      pt_size = unit(3, "mm"),
      set_order = order(ss),
      comb_order = od,
      comb_col = "Purple",
      
      top_annotation = HeatmapAnnotation(
        "DEGs Intersections" =  
          anno_barplot(cs,
                       ylim = c(0, max(cs)*1.1),
                       border = FALSE, 
                       gp = gpar(fill = "#4a452a"), 
                       height = unit(4, "cm")
                       ),
        annotation_name_side = "left",
        annotation_name_rot = 90),
      
      right_annotation = rowAnnotation(
        "No. of DEGs" = 
          anno_barplot(ss, baseline = 500,
                       axis_param = list(
                         at = c(500, 1000, 1500, 2000, 2500),
                         side = "top",
                         labels = c(500, 1000, 1500, 2000, 2500),
                         labels_rot = 45),
                       border = FALSE, 
                       gp = gpar(fill = "#4a452a"), 
                       width = unit(3, "cm")),
        annotation_name_side = "top",
        annotation_name_rot = 0
        ),
      row_names_side = "left", 
      show_row_names = TRUE,
      row_names_gp = grid::gpar(fontsize = 10))

# add numbers of DEGs 
decorate_annotation("DEGs Intersections", {
    grid.text(cs[od], x = seq_along(cs), y = unit(cs[od], "native") + unit(1, "pt"), 
        default.units = "native", just = c("left", "bottom"), 
        gp = gpar(fontsize = 8, col = "#4a452a"), rot = 45)
})
```


```{r Save DEGs results for GO and KEGG analysis}
genes2go %>% saveRDS(file.path("_R", "DEGsCont.rds"))

```




